### PR TITLE
feat: Reflect database state on Files.app

### PR DESCRIPTION
**Pull request by adrien-coye**
### Abstract

FileProvider should reflect the state from UploadQueue and Realm.

### Notes

Superseding https://github.com/Infomaniak/ios-kDrive/pull/1140
Superseding https://github.com/Infomaniak/ios-kDrive/pull/1139

todo: Re-integrate changes from https://github.com/Infomaniak/ios-kDrive/pull/1119

### CODE REVIEW

File: kDriveCore/FileProvider/UploadFileProviderItem.swift [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    public var isUploading: Bool {
+        @InjectService var uploadQueue: UploadQueue
+        let uploadFile = uploadQueue.getUploadingFile(fileProviderItemIdentifier: itemIdentifier.rawValue)
+        return uploadFile != nil
+    }
+
+    public var isUploaded = false
+
+    // TODO: Map error form UploadFile.error
```
Comment (adrien-coye) [inline kDriveCore/FileProvider/UploadFileProviderItem.swift:67]: TODO error

File: kDriveCore/FileProvider/UploadFileProviderItem.swift [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        typeIdentifier = sourceUrl.typeIdentifier ?? UTI.item.identifier
+        capabilities = .allowsAll
+        self.conflictOption = conflictOption
+        self.shouldRemoveAfterUpload = shouldRemoveAfterUpload
+
+        super.init()
+    }
+}
+
+// TODO: Share protocol btwn UploadFileProviderItem / UploadFile
```
Comment (adrien-coye) [inline kDriveCore/FileProvider/UploadFileProviderItem.swift:93]: TODO share protocol

File: kDriveCore/FileProvider/FileProviderItem.swift [hunk 3]
```diff
@@ -161,37 +158,12 @@ public final class FileProviderItem: NSObject, NSFileProviderItem {
...
-        isUploaded = false
-        isDownloaded = false
-        isMostRecentVersionDownloaded = false
-        isShared = false
-        isTrashed = false
-        storageUrl = importedFileUrl
-    }
-
-    public func setUploadingError(_ error: DriveError) {
+// TODO: Refactor this part
```
Comment (adrien-coye) [inline kDriveCore/FileProvider/FileProviderItem.swift:163]: TODO refactor / abstract

File: kDriveCore/Data/UploadQueue/Queue/UploadQueue+Queue.swift [hunk 4]
```diff
@@ -168,6 +173,9 @@ extension UploadQueue: UploadQueueable {
                             addToQueue: Bool = true) -> UploadOperationable? {
         let expiringActivity = ExpiringActivity()
         expiringActivity.start()
+        defer {
+            expiringActivity.endAll()
```
Comment (adrien-coye) [inline kDriveCore/Data/UploadQueue/Queue/UploadQueue+Queue.swift:177]: This fixes the APIV3 branch as a case was not releasing it. a `defer` is more future proof.

File: kDriveCore/FileProvider/UploadFileProviderItem.swift [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        self.userId = userId
+        self.driveId = driveId
+        itemIdentifier = NSFileProviderItemIdentifier(rawValue: "\(uploadFileUUID)")
+        self.sourceUrl = sourceUrl
+        typeIdentifier = sourceUrl.typeIdentifier ?? UTI.item.identifier
+        capabilities = .allowsAll
+        self.conflictOption = conflictOption
+        self.shouldRemoveAfterUpload = shouldRemoveAfterUpload
+
+        if let driveError {
```
Comment (adrien-coye) [inline kDriveCore/FileProvider/UploadFileProviderItem.swift:91]: UploadFile error is now handled, I need to manually check it.

File: kDriveFileProvider/FileProviderExtension+Actions.swift [hunk 6]
```diff
@@ -134,17 +136,21 @@ extension FileProviderExtension {
...
 
-        manager.signalEnumerator(for: parentItemIdentifier) { _ in
-            completionHandler(importedItem, nil)
+        let importItem = UploadFileProviderItem(uploadFileUUID: importedFileUUID,
+                                                parentDirectoryId: parentDirectoryId,
+                                                userId: driveFileManager.drive.userId,
+                                                driveId: driveFileManager.drive.id,
+                                                sourceUrl: storageUrl,
+                                                conflictOption: .version,
+                                                shouldRemoveAfterUpload: false, /* should be true actually ?*/
```
Comment (adrien-coye) [inline kDriveFileProvider/FileProviderExtension+Actions.swift:150]: discuss with @PhilippeWeidmann about `shouldRemoveAfterUpload` shouldn't it be _true_ for FileProvider ?

File: kDriveFileProvider/FileProviderExtension.swift [hunk 7]
```diff
@@ -99,37 +100,59 @@ final class FileProviderExtension: NSFileProviderExtension {
         // Try to reload account if user logged in
         try updateDriveFileManager()
 
+        // TODO: working set in DB
```
Comment (adrien-coye) [inline kDriveFileProvider/FileProviderExtension.swift:103]: TODO: Put a Working Set in database (new one ? )  in a different PR.

File: kDriveFileProvider/FileProviderExtension.swift [hunk 8]
```diff
@@ -99,37 +100,59 @@ final class FileProviderExtension: NSFileProviderExtension {
...
             return item
-        } else if let item = fileProviderState.getImportedDocument(forKey: identifier) {
-            Log.fileProvider("item for identifier - Imported Document")
-            return item
-        } else if let fileId = identifier.toFileId(),
-                  let file = driveFileManager.getCachedFile(id: fileId) {
+        }
+
+        // Read from upload queue
+        // TODO: try with a cache to see if identity of object is used a well as identifier.
```
Comment (adrien-coye) [inline kDriveFileProvider/FileProviderExtension.swift:110]: TODO clean this, message is outdated

File: kDriveFileProvider/FileProviderExtension.swift [hunk 9]
```diff
@@ -187,21 +214,33 @@ final class FileProviderExtension: NSFileProviderExtension {
 
     override func stopProvidingItem(at url: URL) {
         Log.fileProvider("stopProvidingItem at url:\(url)")
-        if let identifier = persistentIdentifierForItem(at: url),
-           let item = try? item(for: identifier) as? FileProviderItem {
+
+        guard let identifier = persistentIdentifierForItem(at: url) else {
+            // The document isn't in realm maybe it was recently imported?
+            // TODO: lookup for a `File` matching
```
Comment (adrien-coye) [inline kDriveFileProvider/FileProviderExtension.swift:220]: TODO: Ask if still needed ?

**Review by adrien-coye — COMMENTED** [review]
Did a self review, also added some annotations to keep track of moving parts.

File: kDriveCore/FileProvider/UploadFileProviderItem.swift (hunk 2 above)
Comment (adrien-coye) [reply kDriveCore/FileProvider/UploadFileProviderItem.swift]: reworked

File: kDriveFileProvider/FileProviderExtension.swift [hunk 10]
```diff
@@ -310,43 +348,27 @@ final class FileProviderExtension: NSFileProviderExtension {
...
-            driveId: driveFileManager.drive.id,
-            fileProviderItemIdentifier: fileProviderItemIdentifier,
-            url: item.storageUrl,
-            name: item.filename,
-            conflictOption: .version,
-            shouldRemoveAfterUpload: false
-        )
+        let uploadFile = uploadFileProviderItem.toUploadFile
 
+        // TODO: Can we remove observation ?
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDriveFileProvider/FileProviderExtension.swift:382]: Keeping observation only for upload completion of specific file for now.

File: kDriveCore/Data/Cache/AccountManager.swift [hunk 11]
```diff
@@ -171,6 +171,7 @@ public class AccountManager: RefreshTokenDelegate, AccountManageable {
         }
     }
 
+    // TODO: Refactor init to explicit DriveContext
```
Comment (adrien-coye) [inline kDriveCore/Data/Cache/AccountManager.swift:174]: Check with @PhilippeWeidmann about `sharedWithMe`

File: kDriveCore/Data/Models/File.swift [hunk 12]
```diff
@@ -740,6 +745,7 @@ public final class File: Object, Codable {
     }
 
     public static func uid(driveId: Int, fileId: Int) -> String {
+        // TODO: Check fix applied: 1_11 collision with 11_1
```
Comment (adrien-coye) [inline kDriveCore/Data/Models/File.swift:748]: Check the fix is applied at next rebase

File: kDriveCore/Data/Models/Upload/UploadFile.swift [hunk 13]
```diff
@@ -330,6 +336,25 @@ public extension UploadFile {
     }
 }
 
+extension UploadFile: FileProviderItemProvider {
+    /// DTO of an UploadFile used by the FileProvider
+    /// Represents an `UploadFile` in the UploadQueue been uploaded
+    public func toFileProviderItem(parent: NSFileProviderItemIdentifier?,
+                                   domain: NSFileProviderDomain?) -> NSFileProviderItem {
+        // TODO: override parent and domain for future working set support.
```
Comment (adrien-coye) [inline kDriveCore/Data/Models/Upload/UploadFile.swift:344]: To be done in the `workingSet` PR

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift [hunk 14]
```diff
@@ -477,39 +480,49 @@ public final class UploadOperation: AsynchronousOperation, UploadOperationable {
...
+        let driveId = readOnlyFile.driveId
+        let userId = readOnlyFile.userId
+
+        let driveFileManagerContext: DriveFileManagerContext
+        @InjectService var appContext: AppContextServiceable
+        switch appContext.context {
+        case .fileProviderExtension:
+            driveFileManagerContext = .fileProvider
+        case .app, .shareExtension, .actionExtension:
+            // TODO: fix for `shared with me`
```
Comment (adrien-coye) [inline kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:495]: Also `share with me` changes

File: kDriveCore/FileProvider/FileProviderItemProvider.swift [hunk 15]
```diff
@@ -0,0 +1,59 @@
+//
```
Comment (adrien-coye) [inline kDriveCore/FileProvider/FileProviderItemProvider.swift:1]: clean

File: kDriveFileProvider/Enumerators/DirectoryEnumerator.swift [hunk 16]
```diff
@@ -42,17 +43,45 @@ final class DirectoryEnumerator: NSObject, NSFileProviderEnumerator {
     func invalidate() {}
 
     func enumerateItems(for observer: NSFileProviderEnumerationObserver, startingAt page: NSFileProviderPage) {
+        let uid = "\(Int.random(in: 0 ... 999))"
```
Comment (adrien-coye) [inline kDriveFileProvider/Enumerators/DirectoryEnumerator.swift:46]: clean

File: kDriveFileProvider/Enumerators/DirectoryEnumerator.swift [hunk 17]
```diff
@@ -144,6 +144,8 @@ final class DirectoryEnumerator: NSObject, NSFileProviderEnumerator {
 
                 var updatedItems = [File]()
                 var deletedItems = [NSFileProviderItemIdentifier]()
+
+                // FIXME: Query upload queue instead
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDriveFileProvider/Enumerators/DirectoryEnumerator.swift:184]: Should just drop it, no longer needed.

File: kDriveFileProvider/FileProviderExtension+Thumbnail.swift [hunk 18]
```diff
@@ -44,14 +44,16 @@ extension FileProviderExtension {
...
 
-            guard file.supportedBy.contains(.thumbnail) else {
-                perThumbnailCompletionHandler(identifier, nil, NSError.featureUnsupported)
-                progress.completedUnitCount += 1
-                if progress.isFinished {
-                    completionHandler(nil)
-                }
-                continue
-            }
+            // TODO: should we just remove this ?
```
Comment (adrien-coye) [inline kDriveFileProvider/FileProviderExtension+Thumbnail.swift:47]: clean

**Review by adrien-coye — COMMENTED** [review]
I did a self review.

File: kDriveCore/Data/Models/Upload/UploadFile.swift [hunk 19]
```diff
@@ -330,6 +332,33 @@ public extension UploadFile {
...
 }
 
+public extension UploadFile {
+    /// Centralise error cleaning
+    func toUploadFileItemProvider() -> UploadFileProviderItem? {
+        guard let pathURL else {
+            return nil
+        }
+
+//        if let cached = Self.cache.object(forKey: id as NSString) {
```
[the remark this answers is not in the document]
Comment (PhilippeWeidmann) [reply kDriveCore/Data/Models/Upload/UploadFile.swift]: Was marked as resolved

File: kDriveCore/Data/Models/Upload/UploadFile.swift [hunk 20]
```diff
@@ -47,6 +47,8 @@ public protocol UploadFilable {
 }
 
 public final class UploadFile: Object, UploadFilable {
+    public static let cache = NSCache<NSString, UploadFileProviderItem>()
```
Comment (PhilippeWeidmann) [inline kDriveCore/Data/Models/Upload/UploadFile.swift]: Also don't forget to remove this part

File: kDriveCore/Data/Cache/AccountManager.swift (hunk 11 above)
Comment (PhilippeWeidmann) [reply kDriveCore/Data/Cache/AccountManager.swift:174]: DriveFileManager with sharedWithMe context is now correctly returned

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift (hunk 14 above)
Comment (PhilippeWeidmann) [reply kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:495]: DriveFileManager with sharedWithMe context is now correctly returned. We should call `instanceWith` only to force fileManager context

File: kDriveCore/FileProvider/FileProviderItem.swift [hunk 21]
```diff
@@ -103,26 +104,46 @@ public final class FileProviderItem: NSObject, NSFileProviderItem {
     public var storageUrl: URL
     public var alreadyEnumerated = false
 
-    public init(file: File, parent: NSFileProviderItemIdentifier? = nil, domain: NSFileProviderDomain?) {
+    // MARK: Static
+
+    public static func getFileName(file: File) -> String {
+        file.name.isEmpty ? "Root" : file.name
```
Comment (PhilippeWeidmann) [inline kDriveCore/FileProvider/FileProviderItem.swift:110]:
```suggestion
        file.name.isEmpty ? "Root" : file.formattedLocalizedName()
```

File: kDriveFileProvider/FileProviderExtension.swift [hunk 22]
```diff
@@ -187,21 +214,33 @@ final class FileProviderExtension: NSFileProviderExtension {
...
+
+        if let item = try? item(for: identifier) as? FileProviderItem {
             if let remoteModificationDate = item.contentModificationDate,
                let localModificationDate = try? item.storageUrl.resourceValues(forKeys: [.contentModificationDateKey])
                .contentModificationDate,
                remoteModificationDate > localModificationDate {
-                backgroundUploadItem(item) {
-                    self.cleanupAt(url: url)
-                }
+                // TODO: check behaviour
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDriveFileProvider/FileProviderExtension.swift:256]: discussed, I change for a simple call to `self.cleanupAt(url: url)` all the time.

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift (hunk 14 above)
Comment (adrien-coye) [reply kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:495]: Ok changed in the last commits, will ask to to check it out.

File: kDriveCore/FileProvider/FileProviderExtensionAdditionalState.swift [hunk 23]
```diff
@@ -24,19 +24,15 @@ import Foundation
 ///
 /// Concrete implementation `FileProviderExtensionAdditionalState` is ThreadSafe
 public protocol FileProviderExtensionAdditionalStatable {
+    // TODO: Remove, read UploadQueue instead
     func importedDocuments(forParent parentItemIdentifier: NSFileProviderItemIdentifier) -> [FileProviderItem]
 
+    // TODO: Remove, read UploadQueue instead
```
Comment (adrien-coye) [inline kDriveCore/FileProvider/FileProviderExtensionAdditionalState.swift:30]: TODO: Clean

File: kDriveCore/FileProvider/FileProviderItemProvider.swift (hunk 15 above)
Comment (adrien-coye) [reply kDriveCore/FileProvider/FileProviderItemProvider.swift:1]: ✨

**Review by adrien-coye — COMMENTED** [review]
Cleaned up all the remaining self remarks and potential cleanups. This should be in a mergeable state.

File: kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift [hunk 24]
```diff
@@ -477,39 +480,48 @@ public final class UploadOperation: AsynchronousOperation, UploadOperationable {
...
-        var parentDirectoryId: Int?
+        let readOnlyFile = try readOnlyFile()
+        let driveId = readOnlyFile.driveId
+        let userId = readOnlyFile.userId
+        @InjectService var appContext: AppContextServiceable
+
+        // Get a DriveFileManager specific to current context
+        let driveFileManager: DriveFileManager?
+        if appContext.context == .fileProviderExtension {
+            driveFileManager = accountManager.getDriveFileManager(for: driveId, userId: userId)?
```
[the remark this answers is not in the document]
Comment (PhilippeWeidmann) [reply kDriveCore/Data/UploadQueue/Operation/UploadOperation.swift:493]: That's correct 👍

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**